src/lib/jws/jwsSigner.js

```diff
@@ -38,50 +38,87 @@ class JwsSigner {
+     *   (see https://github.com/axios/axios)
      */
     sign(requestOptions) {
         this.logger.log(`JWS Signing request: ${util.inspect(requestOptions)}`);
+        const payload = requestOptions.body || requestOptions.data;
+        const uri = requestOptions.uri || requestOptions.url;
 
-        if(!requestOptions.body) {
+        if(!payload) {
             throw new Error('Cannot sign with no body');
         }
 
-        const uriMatches = uriRegex.exec(requestOptions.uri);
+        const uriMatches = uriRegex.exec(uri);
         if(!uriMatches || uriMatches.length < 2) {
-            throw new Error(`URI not valid for protected header: ${requestOptions.uri}`);
+            throw new Error(`URI not valid for protected header: ${uri}`);
         }
 
         // add required JWS headers to the request options
         requestOptions.headers['fspiop-http-method'] = requestOptions.method.toUpperCase();
         requestOptions.headers['fspiop-uri'] = uriMatches[1];
 
+        // get the signature and add it to the header
+        requestOptions.headers['fspiop-signature'] = this.getSignature(requestOptions);
+
+        if(requestOptions.body && typeof(requestOptions.body) !== 'string') {
+            requestOptions.body = JSON.stringify(requestOptions.body);
+        }
+        if(requestOptions.data && typeof(requestOptions.data) !== 'string') {
+            requestOptions.data = JSON.stringify(requestOptions.data);
+        }
+    }
+
+    /**
+     * Returns JWS signature for an outgoing HTTP request options object
+     *
+     * @param requestOptions {object} a request-promise-native/axios style request options object
+     *   (see https://github.com/request/request-promise-native)
+     *   (see https://github.com/axios/axios)
+     *
+     * @returns {string} - JWS Signature as a string 
+    */
+    getSignature(requestOptions) {
+        this.logger.log(`Get JWS Signature: ${util.inspect(requestOptions)}`);
+        const payload = requestOptions.body || requestOptions.data;
+        const uri = requestOptions.uri || requestOptions.url;
+
+        if(!payload) {
+            throw new Error('Cannot sign with no body');
+        }
+
+        const uriMatches = uriRegex.exec(uri);
+        if(!uriMatches || uriMatches.length < 2) {
+            throw new Error(`URI not valid for protected header: ${uri}`);
+        }
+
         // generate the protected header as base64url encoding of UTF-8 encoding of JSON string
 
         // Note: Property names are case sensitive in the protected header object even though they are
         // not case sensitive in the actual HTTP headers
         const protectedHeaderObject = {
             alg: SIGNATURE_ALGORITHM,
             'FSPIOP-URI': requestOptions.headers['fspiop-uri'],
-            'FSPIOP-HTTP-Method': requestOptions.headers['fspiop-http-method'],
+            'FSPIOP-HTTP-Method': requestOptions.method.toUpperCase(),
```

Comment:
Does `requestOptions` always have a `method` property? i.e. do both request-promise-native and axios both use a `method` property to specify the HTTP verb?

Reply:
Yes